From 6af8d85d7a4c4bca50e0c6ff327c129735331cb3 Mon Sep 17 00:00:00 2001
From: Alex Marginean <alexandru.marginean@nxp.com>
Date: Tue, 12 Jun 2018 11:52:19 +0300
Subject: [PATCH 158/706] enetc: pre-set RSS key at PF probe

Signed-off-by: Alex Marginean <alexandru.marginean@nxp.com>
(cherry picked from commit b12fe28e979adb982fd0da6cbeebe1a5cecc608b)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/ethernet/freescale/enetc/enetc.h         |  1 +
 drivers/net/ethernet/freescale/enetc/enetc_ethtool.c | 10 ++++++++--
 drivers/net/ethernet/freescale/enetc/enetc_pf.c      |  5 +++++
 3 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/freescale/enetc/enetc.h b/drivers/net/ethernet/freescale/enetc/enetc.h
index eb399fe69215..aca066a21f4d 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc.h
+++ b/drivers/net/ethernet/freescale/enetc/enetc.h
@@ -244,6 +244,7 @@ void enetc_set_mac_flt_entry(struct enetc_si *si, int index,
 void enetc_clear_mac_flt_entry(struct enetc_si *si, int index);
 int enetc_set_fs_entry(struct enetc_si *si, struct enetc_cmd_rfse *rfse,
 		       int index);
+void enetc_set_rss_key(struct enetc_hw *hw, const u8 *bytes);
 int enetc_set_rss_table(struct enetc_si *si, u16 *table, int len);
 
 #ifdef CONFIG_ENETC_TSN
diff --git a/drivers/net/ethernet/freescale/enetc/enetc_ethtool.c b/drivers/net/ethernet/freescale/enetc/enetc_ethtool.c
index 5bf72abaec0b..b1db0c8178be 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc_ethtool.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc_ethtool.c
@@ -501,6 +501,13 @@ static int enetc_get_rxfh(struct net_device *ndev, u32 *indir, u8 *key,
 			((u32 *)key)[i] = enetc_port_rd(hw, ENETC_PRSSK(i));
 
 	return 0;
+
+void enetc_set_rss_key(struct enetc_hw *hw, const u8 *bytes)
+{
+	int i;
+
+	for (i = 0; i < ENETC_RSSHASH_KEY_SIZE / 4; i++)
+		enetc_port_wr(hw, ENETC_PRSSK(i), ((u32 *)bytes)[i]);
 }
 
 static int enetc_set_rxfh(struct net_device *ndev, const u32 *indir,
@@ -513,8 +520,7 @@ static int enetc_set_rxfh(struct net_device *ndev, const u32 *indir,
 
 	/* set hash key, if PF */
 	if (key && hw->port)
-		for (i = 0; i < ENETC_RSSHASH_KEY_SIZE / 4; i++)
-			enetc_port_wr(hw, ENETC_PRSSK(i), ((u32 *)key)[i]);
+		enetc_set_rss_key(hw, key);
 
 	/* set RSS table */
 	if (indir) {
diff --git a/drivers/net/ethernet/freescale/enetc/enetc_pf.c b/drivers/net/ethernet/freescale/enetc/enetc_pf.c
index 3c2a5540b02b..ff4bcca73a50 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc_pf.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc_pf.c
@@ -534,12 +534,17 @@ static void enetc_configure_port_mac(struct enetc_hw *hw)
 
 static void enetc_configure_port(struct enetc_pf *pf)
 {
+	u8 hash_key[ENETC_RSSHASH_KEY_SIZE];
 	struct enetc_hw *hw = &pf->si->hw;
 
 	enetc_configure_port_mac(hw);
 
 	enetc_port_si_configure(pf->si);
 
+	/* set up hash key */
+	get_random_bytes(hash_key, ENETC_RSSHASH_KEY_SIZE);
+	enetc_set_rss_key(hw, hash_key);
+
 	/* split up RFS entries */
 	enetc_port_alloc_rfs(pf->si);
 
-- 
2.17.1

